Привет, меня зовут Люман Аблаев. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>


### <font color='orange'>Общее впечатление</font>
* Спасибо за то, что прислал такую хорошую работу, видно, что было приложено много усилий.
* Выводы  описаны грамотно и логично. Они просты для понимания.
* В целом проект очень аккуратный, приятно проверять такие работы.
* Я оставил некоторые советы, обрати на них внимание.
* Нужно поправить момент с доверительным интервалом. Я думаю ты легко с ними справишься.
* Жду твою обновленную работу.

### <font color='orange'>Общее впечатление[2]</font>
* Спасибо за очень оперативные правки своей работы.
* Радует, что ты реагируешь на мои комментарии.
* Теперь твой проект выглядит законченным. Поэтому я его принимаю и желаю удачи в дальнейшем обучении.
* Продолжай в том же духе! =)

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> 
    
Привет Люман, спасибо за проверку! я тогда буду оставлять комментарии в таких блоках.

</div>

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Общее-впечатление" data-toc-modified-id="Общее-впечатление-0.0.1"><span class="toc-item-num">0.0.1&nbsp;&nbsp;</span><font color="orange">Общее впечатление</font></a></span></li><li><span><a href="#Общее-впечатление[2]" data-toc-modified-id="Общее-впечатление[2]-0.0.2"><span class="toc-item-num">0.0.2&nbsp;&nbsp;</span><font color="orange">Общее впечатление[2]</font></a></span></li></ul></li></ul></li><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Задача" data-toc-modified-id="Задача-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Задача</a></span></li><li><span><a href="#Загрузка-и-подготовка-данных" data-toc-modified-id="Загрузка-и-подготовка-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Загрузка и подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Описание данных</a></span><ul class="toc-item"><li><span><a href="#Регион-№1" data-toc-modified-id="Регион-№1-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Регион №1</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-региону-1" data-toc-modified-id="Вывод-по-региону-1-3.1.1.1"><span class="toc-item-num">3.1.1.1&nbsp;&nbsp;</span>Вывод по региону 1</a></span></li></ul></li><li><span><a href="#Регион-№2" data-toc-modified-id="Регион-№2-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Регион №2</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-региону-2" data-toc-modified-id="Вывод-по-региону-2-3.1.2.1"><span class="toc-item-num">3.1.2.1&nbsp;&nbsp;</span>Вывод по региону 2</a></span></li></ul></li><li><span><a href="#Регион-№3" data-toc-modified-id="Регион-№3-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Регион №3</a></span><ul class="toc-item"><li><span><a href="#Вывод-по-региону-3" data-toc-modified-id="Вывод-по-региону-3-3.1.3.1"><span class="toc-item-num">3.1.3.1&nbsp;&nbsp;</span>Вывод по региону 3</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Обучение-и-проверка-модели" data-toc-modified-id="Обучение-и-проверка-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Обучение и проверка модели</a></span><ul class="toc-item"><li><span><a href="#Регион-№1" data-toc-modified-id="Регион-№1-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Регион №1</a></span></li><li><span><a href="#Регион-№2" data-toc-modified-id="Регион-№2-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Регион №2</a></span></li><li><span><a href="#Регион-№3" data-toc-modified-id="Регион-№3-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Регион №3</a></span></li><li><span><a href="#Вывод-по-блоку" data-toc-modified-id="Вывод-по-блоку-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Вывод по блоку</a></span></li></ul></li><li><span><a href="#3.-Подготовка-к-расчёту-прибыли" data-toc-modified-id="3.-Подготовка-к-расчёту-прибыли-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>3. Подготовка к расчёту прибыли</a></span><ul class="toc-item"><li><span><a href="#Основные-данные" data-toc-modified-id="Основные-данные-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Основные данные</a></span><ul class="toc-item"><li><span><a href="#Регион-№1" data-toc-modified-id="Регион-№1-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>Регион №1</a></span></li><li><span><a href="#Регион-№2" data-toc-modified-id="Регион-№2-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>Регион №2</a></span></li><li><span><a href="#Регион-№3" data-toc-modified-id="Регион-№3-5.1.3"><span class="toc-item-num">5.1.3&nbsp;&nbsp;</span>Регион №3</a></span></li></ul></li><li><span><a href="#Вывод-по-блоку" data-toc-modified-id="Вывод-по-блоку-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Вывод по блоку</a></span></li></ul></li><li><span><a href="#4.-Расчёт-прибыли-и-рисков" data-toc-modified-id="4.-Расчёт-прибыли-и-рисков-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>4. Расчёт прибыли и рисков</a></span><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Регион-№1" data-toc-modified-id="Регион-№1-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>Регион №1</a></span></li><li><span><a href="#Регион-№2" data-toc-modified-id="Регион-№2-6.1.2"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>Регион №2</a></span></li><li><span><a href="#Регион-№3" data-toc-modified-id="Регион-№3-6.1.3"><span class="toc-item-num">6.1.3&nbsp;&nbsp;</span>Регион №3</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

In [1]:
import sklearn
sklearn.__version__

'0.21.2'

# Описание проекта

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

# Задача
1. **Построить модель для определения региона, где добыча принесёт наибольшую прибыль.**
2. **Проанализировать возможную прибыль и риски.**

<div class="alert alert-block alert-success">
<b>Успех:</b> Спасибо за подробное описание проекта.
</div>


# Загрузка и подготовка данных

In [2]:
# импорт данных 
import os
import errno
from urllib.request import urlretrieve
from IPython.display import HTML, display
from pathlib import Path
import urllib


# Основные библиотеки для работы с данными
import pandas as pd
import numpy as np
import math

# работа с графиками
from matplotlib import pyplot as plt
import scipy
import seaborn as sns
from sklearn import metrics

# библиотека для модели
from sklearn.linear_model import LinearRegression

# Настройка моделей и параметров (разбивка датасета, его перемешивание и т.д.)
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from tqdm import tqdm
from itertools import product 
from sklearn.metrics import f1_score, make_scorer
from sklearn.utils import shuffle

# Наши основные показатели
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve 
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error


from scipy import stats as st
from scipy.stats import t
from numpy.random import RandomState
import urllib
import urllib.request


In [3]:
# отключение предупреждений
import warnings; warnings.filterwarnings("ignore", category=Warning)

In [4]:
# обновление библиотек Numpy и Pandas для исключения ошибок версий
# !pip3 install --upgrade --user --quiet --no-warn-script-location numpy==1.19.3 pandas

In [5]:
# для просмотра всех столбцов таблицы
pd.options.display.max_columns = None

In [6]:
# # обновление библиотек
# !pip3 install --upgrade --user --quiet matplotlib seaborn

In [7]:
# # обновление библиотек
# !pip3 install --upgrade --user --quiet sklearn

In [8]:
def load_from_url(file_url, file_path):
    """Загрузка файла из сетевого источника и загрузка в file_path"""
    folder_path = os.path.dirname(file_path)
    if not os.path.exists(folder_path):
        print('Создаем папку {folder_path}')
        os.makedirs(folder_path)
        print(f'Закгружаю файл из "{file_url}" в "{file_path}" ... ', end='')
        result = urllib.urlretrieve(url=file_url, filename=file_path) 
        print('OK')
        return result



In [9]:
def get_data_frame(file_path, *args, **kwqrgs):
    """Загружаю файл из локального файла file_path"""
    print(f'Открываю "{file_path}" с помошью Pandas ...', end='')
    result = pd.read_csv(file_path, *args, **kwqrgs)
    print('OK')
    return result
    

In [10]:
# 
def load_dataframe(file_url, file_path, *args, **kwqrgs):
    if os.path.isfile(file_path):
        print(f'Файл "{file_path}" уже был ранее заргужен.')
    else:
        load_from_url(file_url, file_path)
#     Создаем датафрейм
    return get_data_frame(file_path, *args, **kwqrgs)
    

<div class="alert alert-block alert-success">
<b>Успех:</b>Молодец, что собираешь все импорты, настройки, и вспомогательные функции в начале проекта.
</div>


In [11]:
dataset_url0 = '.../datasets/geo_data_0.csv'
dataset_url1 = '.../datasets/geo_data_1.csv'
dataset_url2 = '.../datasets/geo_data_2.csv'


# dataset_path0 = '/Users/Matvey/projects/geo_data_0.csv'
# dataset_path1 = '/Users/Matvey/projects/geo_data_1.csv'
# dataset_path2 = '/Users/Matvey/projects/geo_data_2.csv'

dataset_path0 = 'D:/projects/geo_data_0.csv'
dataset_path1 = 'D:/projects/geo_data_1.csv'
dataset_path2 = 'D:/projects/geo_data_2.csv'

In [12]:
# raw_data0 = load_dataframe(dataset_url0, dataset_path0)
# raw_data1 = load_dataframe(dataset_url1, dataset_path1)
# raw_data2 = load_dataframe(dataset_url2, dataset_path2)

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> 
    
Я вносил правки не со своего компьтера, поэтому ниже просто ипортировал файлы по ссылкам, а верхний блоек закомментировал.

</div>

In [13]:
raw_data0 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_0.csv')
raw_data1 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_1.csv')
raw_data2 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_2.csv')

## Описание данных
- id — уникальный идентификатор скважины;
- f0, f1, f2 — три признака точек;
- product — объём запасов в скважине (тыс. баррелей).

### Регион №1

In [14]:
raw_data0

id        f0        f1        f2     product
0      txEyH  0.705745 -0.497823  1.221170  105.280062
1      2acmU  1.334711 -0.340164  4.365080   73.037750
2      409Wp  1.022732  0.151990  1.419926   85.265647
3      iJLyR -0.032172  0.139033  2.978566  168.620776
4      Xdl7t  1.988431  0.155413  4.751769  154.036647
...      ...       ...       ...       ...         ...
99995  DLsed  0.971957  0.370953  6.075346  110.744026
99996  QKivN  1.392429 -0.382606  1.273912  122.346843
99997  3rnvd  1.029585  0.018787 -1.348308   64.375443
99998  7kl59  0.998163 -0.528582  1.583869   74.040764
99999  1CWhH  1.764754 -0.266417  5.722849  149.633246

[100000 rows x 5 columns]

In [15]:
raw_data0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [16]:
raw_data0.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

In [17]:
raw_data0.corr()

f0        f1        f2   product
f0       1.000000 -0.440723 -0.003153  0.143536
f1      -0.440723  1.000000  0.001724 -0.192356
f2      -0.003153  0.001724  1.000000  0.483663
product  0.143536 -0.192356  0.483663  1.000000

In [18]:
raw_data0.duplicated().sum()

0

#### Вывод по региону 1

Данные нормальные. Нет пропусков и дубликатов. Также не видно сильных аномалий

### Регион №2

In [19]:
raw_data1

id         f0         f1        f2     product
0      kBEdx -15.001348  -8.276000 -0.005876    3.179103
1      62mP7  14.272088  -3.475083  0.999183   26.953261
2      vyE1P   6.263187  -5.948386  5.001160  134.766305
3      KcrkZ -13.081196 -11.506057  4.999415  137.945408
4      AHL4O  12.702195  -8.147433  5.004363  134.766305
...      ...        ...        ...       ...         ...
99995  QywKC   9.535637  -6.878139  1.998296   53.906522
99996  ptvty -10.160631 -12.558096  5.005581  137.945408
99997  09gWa  -7.378891  -3.084104  4.998651  137.945408
99998  rqwUm   0.665714  -6.152593  1.000146   30.132364
99999  relB0  -3.426139  -7.794274 -0.003299    3.179103

[100000 rows x 5 columns]

In [20]:
raw_data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [21]:
raw_data1.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

In [22]:
raw_data1.corr()

f0        f1        f2   product
f0       1.000000  0.182287 -0.001777 -0.030491
f1       0.182287  1.000000 -0.002595 -0.010155
f2      -0.001777 -0.002595  1.000000  0.999397
product -0.030491 -0.010155  0.999397  1.000000

In [23]:
raw_data1.duplicated().sum()

0

#### Вывод по региону 2

Данные нормальные. Нет пропусков и дубликатов. Не видно сильных аномалий.

Также есть интересная особенность параметр f2 очень сильно коррелирует с цеелвым признаком. Далее мы попробуем только на этом признаке обучить модель, посмотрим что из этого получится.

### Регион №3


In [24]:
raw_data2

id        f0        f1        f2     product
0      fwXo0 -1.146987  0.963328 -0.828965   27.758673
1      WJtFt  0.262778  0.269839 -2.530187   56.069697
2      ovLUW  0.194587  0.289035 -5.586433   62.871910
3      q6cA6  2.236060 -0.553760  0.930038  114.572842
4      WPMUX -0.515993  1.716266  5.899011  149.600746
...      ...       ...       ...       ...         ...
99995  4GxBu -1.777037  1.125220  6.263374  172.327046
99996  YKFjq -1.261523 -0.894828  2.524545  138.748846
99997  tKPY3 -1.199934 -2.957637  5.219411  157.080080
99998  nmxp2 -2.419896  2.417221 -5.548444   51.795253
99999  V9kWn -2.551421 -2.025625  6.090891  102.775767

[100000 rows x 5 columns]

In [25]:
raw_data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [26]:
raw_data2.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

In [27]:
raw_data2.corr()

f0        f1        f2   product
f0       1.000000  0.000528 -0.000448 -0.001987
f1       0.000528  1.000000  0.000779 -0.001012
f2      -0.000448  0.000779  1.000000  0.445871
product -0.001987 -0.001012  0.445871  1.000000

In [28]:
raw_data2.duplicated().sum()

0

#### Вывод по региону 3

Данные нормальные. Нет пропусков и дубликатов. Не видно сильных аномалий.

<div class="alert alert-block alert-success">
<b>Успех:</b> Загрузка данных и первичный анализ сделаны корректно
</div>


# Обучение и проверка модели

In [29]:
# Зададим значения randomState для получения одинаковых результатов
randomState = 42
state = RandomState(randomState) 

## Регион №1

In [30]:
# Выделяем отдельно признаки, отдельно целевой признак
# Столбец id нам также не нужен для обучения
features0 = raw_data0.drop(['product', 'id'], axis=1) 
target0 = raw_data0['product']
# features0

In [31]:
# Делим датаест на обучающую и валидационную выборки
features_train0, features_valid0, target_train0, target_valid0 = train_test_split(
    features0, target0, test_size=0.25, random_state=randomState)

In [32]:
# Наша модель которая будет делать все предсказания
model = LinearRegression()

In [33]:
# Функция вывода всей нужной нам информации о модели
def model_specifications(features_train, features_valid, target_train, target_valid):
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)
    result = mean_squared_error(target_valid, predictions_valid)**0.5 # RMSE на валидационной выборке
    print("RMSE модели линейной регрессии на валидационной выборке: {:.2f}".format(result))

    raw_material_stock_pred = predictions_valid.mean()
    raw_material_stock_real = target_valid.mean()
    print('Средний запас предсказанного сырья - {:.2f}'.format(raw_material_stock_pred))
    print('Средний запас реального сырья - {:.2f}'.format(raw_material_stock_real))
    confidence_interval = t.interval(0.95, len(target_valid0), target_valid0.mean(), target_valid0.sem())
    print('95%-ый доверительный интервал:', confidence_interval)


In [34]:
model_specifications(features_train0, features_valid0, target_train0, target_valid0)

RMSE модели линейной регрессии на валидационной выборке: 37.76
Средний запас предсказанного сырья - 92.40
Средний запас реального сырья - 92.33
95%-ый доверительный интервал: (91.77707015145042, 92.87484259023732)


Попробуем нормализовать числовые данные, может это позволит уменьшить RMSE

In [35]:
features = ['f0', 'f1', 'f2']
scaler = StandardScaler()
scaler.fit(features0[features])
features0[features] = scaler.transform(features0[features])
features0

f0        f1        f2
0      0.235512 -1.482793 -0.394515
1      0.956946 -1.170246  0.573368
2      0.599101 -0.194581 -0.333326
3     -0.610890 -0.220269  0.146516
4      1.706773 -0.187797  0.692414
...         ...       ...       ...
99995  0.540861  0.239498  1.099890
99996  1.023149 -1.254384 -0.378278
99997  0.606961 -0.458648 -1.185554
99998  0.570920 -1.543771 -0.282855
99999  1.450212 -1.024046  0.991371

[100000 rows x 3 columns]

<div class="alert alert-block alert-info">
<b>Совет:</b> Упс.... здесь небольшая ошибка, не критичная, но важная.<br>
Масштабировать признаки на всем датасете может привести к утечке данных (data leak). Обучать scaler нужно только на трейне.
Про Data Leaks можно почитать подробнее по следующим ссылкам:

[Ссылка 1](https://towardsdatascience.com/preventing-data-leakage-in-your-machine-learning-model-9ae54b3cd1fb) <br>
[Ссылка 2](https://machinelearningmastery.com/data-preparation-without-data-leakage/)
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> 
    
Понял, спасибо, обязательно посотрю!
    
Ну там ниже я пишу что мы не используем нормализованные данные. 
</div>

In [36]:
# Еще раз разбиваем наш датасет
features_train0, features_valid0, target_train0, target_valid0 = train_test_split(
    features0, target0, test_size=0.25, random_state=randomState)


In [37]:
# Сбрасываем индексы, это нам понадобитс для дальнейшего бутстрепа
target_valid0.reset_index(drop=True, inplace=True)


In [38]:
# И повторно запускаем нашу функцию
model_specifications(features_train0, features_valid0, target_train0, target_valid0)

RMSE модели линейной регрессии на валидационной выборке: 37.76
Средний запас предсказанного сырья - 92.40
Средний запас реального сырья - 92.33
95%-ый доверительный интервал: (91.77707015145042, 92.87484259023732)


**Нормализация числовых переменных не дала никакого результата, в бальнейшем делать этого не будем**

**Для удобства запишем все в таблицу**

In [39]:
results = pd.DataFrame(index=['RMSE на валидационной выборке', 'Средний запас предсказанного сырья', 
                              'Средний запас реального сырья'])
results['Сырые денные регион 1'] = ['37.76', '92.40', '92.33']
results

Сырые денные регион 1
RMSE на валидационной выборке                      37.76
Средний запас предсказанного сырья                 92.40
Средний запас реального сырья                      92.33

In [40]:
# confidence_interval = t.interval(0.95, len(target_valid0), target_valid0.mean(), target_valid0.sem())
# confidence_interval

## Регион №2

In [41]:
features1 = raw_data1.drop(['product', 'id'], axis=1) 
target1 = raw_data1['product']
# features1


In [42]:
features_train1, features_valid1, target_train1, target_valid1 = train_test_split(
    features1, target1, test_size=0.25, random_state=randomState)

target_valid1.reset_index(drop=True, inplace=True)

In [43]:
model_specifications(features_train1, features_valid1, target_train1, target_valid1)

RMSE модели линейной регрессии на валидационной выборке: 0.89
Средний запас предсказанного сырья - 68.71
Средний запас реального сырья - 68.73
95%-ый доверительный интервал: (91.77707015145042, 92.87484259023732)


Также напомню что у нас странным образом коррелировали столбец со значением f2 и "product" — объём запасов в скважине. Попробуем убрать признаки 'f0', 'f1' и обучить модель без них. 

In [44]:
features1_1 = raw_data1.drop(['product', 'id', 'f0', 'f1'], axis=1) 

In [45]:
features_train1_1, features_valid1_1, target_train1_1, target_valid1_1 = train_test_split(
    features1_1, target1, test_size=0.25, random_state=randomState)
# features_train1


In [46]:
model_specifications(features_train1_1, features_valid1_1, target_train1_1, target_valid1_1)

RMSE модели линейной регрессии на валидационной выборке: 1.60
Средний запас предсказанного сырья - 68.72
Средний запас реального сырья - 68.73
95%-ый доверительный интервал: (91.77707015145042, 92.87484259023732)


Результат получился плачевный, отклонение стало только больше. 

Вернем признаки в датасет по первому региону т.к. на этих данных модель обучилась лучше

In [47]:
# features1 = raw_data1.drop(['product', 'id'], axis=1) 

In [48]:
results['Сырые денные регион 2'] = [' 0.89', '68.71', '68.73']
results

Сырые денные регион 1 Сырые денные регион 2
RMSE на валидационной выборке                      37.76                  0.89
Средний запас предсказанного сырья                 92.40                 68.71
Средний запас реального сырья                      92.33                 68.73

## Регион №3


In [49]:
features2 = raw_data2.drop(['product', 'id'], axis=1) 
target2 = raw_data2['product']
# features1

In [50]:
features_train2, features_valid2, target_train2, target_valid2 = train_test_split(
    features2, target2, test_size=0.25, random_state=randomState)

target_valid2.reset_index(drop=True, inplace=True)

In [51]:
model_specifications(features_train2, features_valid2, target_train2, target_valid2)

RMSE модели линейной регрессии на валидационной выборке: 40.15
Средний запас предсказанного сырья - 94.77
Средний запас реального сырья - 95.15
95%-ый доверительный интервал: (91.77707015145042, 92.87484259023732)


In [52]:
results['Сырые денные регион 3'] = ['40.15', '94.77', '95.15']
results

Сырые денные регион 1  \
RMSE на валидационной выборке                      37.76   
Средний запас предсказанного сырья                 92.40   
Средний запас реального сырья                      92.33   

                                   Сырые денные регион 2 Сырые денные регион 3  
RMSE на валидационной выборке                       0.89                 40.15  
Средний запас предсказанного сырья                 68.71                 94.77  
Средний запас реального сырья                      68.73                 95.15

## Вывод по блоку

Мы получили 3 основных значения:
- RMSE на валидационной выборке **это корень из среднеквадратичной ошибки наших прогнозов**
- Средний запас предсказанного сырья
- Средний запас реального сырья	

Мы видимо что во 2-м регионе у нас отклонение от реальных значений самое маленькое, значит тут наша модель сработала лучше всего.
Также видно что средний запас предсказанного сырья не сильно отличается от реального запаса, но средние значения нам очень трудно интерпретировать. С учетом того что мы знает RMSE и видим что в 1-м и 3-м регионе отклонение может достигать почти половины от реальных значений.

Для того чтобы определить наилучший регион для разработок, проведем исследование наших данных техникой Bootstrap. 

<div class="alert alert-block alert-success">
<b>Успех:</b> Шаг сделан хороо, все метрики посчитаны корректно.
</div>


# 3. Подготовка к расчёту прибыли

## Основные данные

- Бюджет на разработку скважин в регионе — 10 млрд рублей. (budget - Development budget)
- Доход с каждой единицы продукта составляет 450 тыс. рублей (barel - Price per barrel * 1000)

In [53]:
budget = 10000000 # считаем в тысячах, поэтому получаем 10 миллионов тысяч
well = 200 #Количество скважин
budget_one = budget / well
barel = 450 #это в тысячах (450 000)

# Well budget - wbudget Бюджет на одну скважину (из расчета что в регионе будет всего 200 скважин)
# quantity of oil - oilq - минимальный объём запасов в скважине для того чтобы проетк был прибыльный

<div class="alert alert-block alert-info">
<b>Совет:</b> Константы принятно записывать в ВЕРХНЕМ регистре=)
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> 
    
Ок, я почему-то думал что это касается только совсем не меняющихся значений)
</div>

In [54]:
wbudget = budget / 200 #Бюджет на бурение одной скважины
oilq = wbudget / 450
print('Минимальный объём запасов в скважине нужный для того чтобы скважина считалась удачной'
      '- {:.2f} тысяч бырелей'.format(oilq))

Минимальный объём запасов в скважине нужный для того чтобы скважина считалась удачной- 111.11 тысяч бырелей


### Регион №1

In [55]:
model.fit(features_train0, target_train0)
predictions_valid0 = model.predict(features_valid0)
result0 = mean_squared_error(target_valid0, predictions_valid0)**0.5 # RMSE на валидационной выборке
# print("RMSE модели линейной регрессии на валидационной выборке: {:.2f}".format(result0))

raw_material_stock_pred0 = predictions_valid0.mean()
raw_material_stock_real0 = target_valid0.mean()
# print('Средний запас предсказанного сырья - {:.2f}'.format(raw_material_stock_pred0))
# print('Средний запас реального сырья - {:.2f}'.format(raw_material_stock_real0))
confidence_interval0 = t.interval(0.95, len(target_valid0), target_valid0.mean(), target_valid0.sem())
# print('95%-ый доверительный интервал: {:.2f} : {:.2f}'.format(confidence_interval0[0], confidence_interval0[1]))

**Для удобства заносим все в таблицу**

In [56]:
results2 = pd.DataFrame(index=['RMSE модели линейной регрессии на валидационной выборке', 'Средний запас предсказанного сырья', 
                              'Средний запас реального сырья', '95%-ый доверительный интервал'])
results2['Данные предсказаний регион 1'] = [result0, raw_material_stock_pred0, 
                                            raw_material_stock_real0, 
                                            '{:.2f} : {:.2f}'.format(confidence_interval0[0], confidence_interval0[1])]
results2

Данные предсказаний регион 1
RMSE модели линейной регрессии на валидационной...                      37.7566
Средний запас предсказанного сырья                                      92.3988
Средний запас реального сырья                                            92.326
95%-ый доверительный интервал                                     91.78 : 92.87

### Регион №2

In [57]:
model.fit(features_train1, target_train1)
predictions_valid1 = model.predict(features_valid1)
result1 = mean_squared_error(target_valid1, predictions_valid1)**0.5 # RMSE на валидационной выборке
# print("RMSE модели линейной регрессии на валидационной выборке: {:.2f}".format(result1))

raw_material_stock_pred1 = predictions_valid1.mean()
raw_material_stock_real1 = target_valid1.mean()
# print('Средний запас предсказанного сырья - {:.2f}'.format(raw_material_stock_pred1))
# print('Средний запас реального сырья - {:.2f}'.format(raw_material_stock_real1))
confidence_interval1 = t.interval(0.95, len(target_valid1), target_valid1.mean(), target_valid1.sem())
# print('95%-ый доверительный интервал:', confidence_interval1)

In [58]:
results2['Данные предсказаний регион 2'] = [result1, raw_material_stock_pred1, 
                                            raw_material_stock_real1, '{:.2f} : {:.2f}'.format(confidence_interval1[0], 
                                                                                               confidence_interval1[1])]
results2

Данные предсказаний регион 1  \
RMSE модели линейной регрессии на валидационной...                      37.7566   
Средний запас предсказанного сырья                                      92.3988   
Средний запас реального сырья                                            92.326   
95%-ый доверительный интервал                                     91.78 : 92.87   

                                                   Данные предсказаний регион 2  
RMSE модели линейной регрессии на валидационной...                      0.89028  
Средний запас предсказанного сырья                                      68.7129  
Средний запас реального сырья                                           68.7254  
95%-ый доверительный интервал                                     68.16 : 69.29

### Регион №3

In [59]:
model.fit(features_train2, target_train2)
predictions_valid2 = model.predict(features_valid2)
result2 = mean_squared_error(target_valid1, predictions_valid2)**0.5 # RMSE на валидационной выборке
# print("RMSE модели линейной регрессии на валидационной выборке: {:.2f}".format(result2))

raw_material_stock_pred2 = predictions_valid2.mean()
raw_material_stock_real2 = target_valid2.mean()
# print('Средний запас предсказанного сырья - {:.2f}'.format(raw_material_stock_pred2))
# print('Средний запас реального сырья - {:.2f}'.format(raw_material_stock_real2))
confidence_interval2 = t.interval(0.95, len(target_valid2), target_valid2.mean(), target_valid2.sem())
# print('95%-ый доверительный интервал:', confidence_interval2)

In [60]:
results2['Данные предсказаний регион 3'] = [result2, raw_material_stock_pred2, 
                                            raw_material_stock_real2, '{:.2f} : {:.2f}'.format(confidence_interval2[0], confidence_interval2[1])]
results2

Данные предсказаний регион 1  \
RMSE модели линейной регрессии на валидационной...                      37.7566   
Средний запас предсказанного сырья                                      92.3988   
Средний запас реального сырья                                            92.326   
95%-ый доверительный интервал                                     91.78 : 92.87   

                                                   Данные предсказаний регион 2  \
RMSE модели линейной регрессии на валидационной...                      0.89028   
Средний запас предсказанного сырья                                      68.7129   
Средний запас реального сырья                                           68.7254   
95%-ый доверительный интервал                                     68.16 : 69.29   

                                                   Данные предсказаний регион 3  
RMSE модели линейной регрессии на валидационной...                      56.4472  
Средний запас предсказанного сырья                                       94.771  
Средний запас реального сырья                                            95.151  
95%-ый доверительный интервал                                     94.60 : 95.71

Как мы видим в регионе номер 2 у нас самое маленькое RMSE, это значит что данные у нас там распределены с меньшим разбросом. 
При этом средний запас сырья там меньше всего. 

## Вывод по блоку

Здесь мы подготовили данные, а именно все ключевые значения для расчётов сохранили в отдельных переменных.

Также мы знаем, что минимальный объём запасов в скважине нужный для того чтобы скважина считалась безубыточный должен составлять - 111.11 тысяч бырелей. Здесь же мы видим что ни один из регионов по этому параметру не подходит. Но нужно не забывать что мы рассматриваем массим из 25 000 скважин. Знаит там есть как хорошие, так и плохие. 

Наша основаная задача теперь состоит в том чтобы через Bootstrap найти эти скважины и посчитать все риски.

<div class="alert alert-block alert-success">
<b>Успех:</b> Значение для безубыточной разработки посчитано верно, выводы логичны.
</div>

# 4. Расчёт прибыли и рисков 

## Подготовка данных

In [61]:
# переводим оббъекты в Series
predictions_valid0 = pd.Series(predictions_valid0) 
predictions_valid1 = pd.Series(predictions_valid1)
predictions_valid2 = pd.Series(predictions_valid2)

In [62]:
# Напишем функцию расчета прибыли

def profit(pred_valid, target_valid, n):
    predictions_valid_sorted = pred_valid.sort_values(ascending=False)
    target_valid_selected = target_valid[predictions_valid_sorted.index][:n]
    result = target_valid_selected.sum() * barel - budget
#     print('Прибыль с 200 лучших скважин составила - {:.2f} тысяч рублей'.format(result))
    return result

# profit(predictions_valid0, predictions_valid0, 200) # для проверки

### Регион №1

In [63]:
# Все значения выручки из нашей выбоки будут храниться в переменной values0 (для каждого региона своя переменная)
values0 = []

# Производим бутстреп
for i in range(1000):
    target_valid_sub0 = target_valid0.sample(n=500, replace=True, random_state=state)
    pred_valid_subsample0 = predictions_valid0[target_valid_sub0.index]
    predictions_valid_sorted0 = pred_valid_subsample0.sort_values(ascending=False)
    values0.append(profit(predictions_valid_sorted0, target_valid_sub0, 200)) #Добавление значений
    
values0 = pd.Series(values0)

mean0 = values0.mean()
print("Средняя прибыль: {:.2f}".format(mean0))

lower0 = values0.quantile(.05)
upper0 = values0.quantile(.95)

# # Определяем 95%-й доверительный интервал
# interval0 = t.interval(0.95, len(values0), values0.mean(), values0.sem())

# print(interval0)
print('95% доверительный интервал в регионе (тысяч руб.) - {:.2f} - {:.2f}'.format(lower0, upper0))

Средняя прибыль: 427847.56
95% доверительный интервал в регионе (тысяч руб.) - -9708.12 - 865127.64


In [64]:
# Расчитываем риск убытков
risk_of_loss0 = values0[values0 < 0].count()/1000
print('Риск убытков в регионе 1: {:.2%}'.format(risk_of_loss0))

Риск убытков в регионе 1: 5.50%


**Для удобства запишем все в таблицу**

In [65]:
results3 = pd.DataFrame(index=['Средняя прибыль в регионе после Bootstrap (тысяч руб.)', 
                               '95% доверительный интервал в регионе (тысяч руб.)', 
                              'Риск убытков в регионе в %'])
results3['Данные предсказаний регион 1'] = ['{:.2f}'.format(mean0), 
                                            '{:.2f} : {:.2f}'.format(lower0, upper0), '{:.2%}'.format(risk_of_loss0)]
results3

Данные предсказаний регион 1
Средняя прибыль в регионе после Bootstrap (тыся...                    427847.56
95% доверительный интервал в регионе (тысяч руб.)          -9708.12 : 865127.64
Риск убытков в регионе в %                                                5.50%

### Регион №2

In [66]:
values1 = []

for i in range(1000):
    target_valid_sub1 = target_valid1.sample(n=500, replace=True, random_state=state)
    pred_valid_subsample1 = predictions_valid1[target_valid_sub1.index]
    predictions_valid_sorted1 = pred_valid_subsample1.sort_values(ascending=False)
    values1.append(profit(predictions_valid_sorted1, target_valid_sub1, 200))
      
values1 = pd.Series(values1)

mean1 = values1.mean()
print("Средняя прибыль: {:.2f}".format(mean1))

# interval1 = t.interval(0.95, len(values1), values1.mean(), values1.sem())

# print(interval1)
lower1 = values1.quantile(.05)
upper1 = values1.quantile(.95)

print('95% доверительный интервал в регионе (тысяч руб.): {:.2f} - {:.2f}'.format(lower1, upper1))


Средняя прибыль: 511530.22
95% доверительный интервал в регионе (тысяч руб.): 149485.76 - 862955.17


In [67]:
risk_of_loss1 = values0[values1 < 0].count()/1000
print('Риск убытков в регионе 1: {:.2%}'.format(risk_of_loss1))


Риск убытков в регионе 1: 0.60%


In [68]:
results3['Данные предсказаний регион 2'] = ['{:.2f}'.format(mean1), 
                                            '{:.2f} : {:.2f}'.format(lower1, upper1), '{:.2%}'.format(risk_of_loss1)]
results3

Данные предсказаний регион 1  \
Средняя прибыль в регионе после Bootstrap (тыся...                    427847.56   
95% доверительный интервал в регионе (тысяч руб.)          -9708.12 : 865127.64   
Риск убытков в регионе в %                                                5.50%   

                                                   Данные предсказаний регион 2  
Средняя прибыль в регионе после Bootstrap (тыся...                    511530.22  
95% доверительный интервал в регионе (тысяч руб.)         149485.76 : 862955.17  
Риск убытков в регионе в %                                                0.60%

### Регион №3

In [69]:
values2 = []

for i in range(1000):
    target_valid_sub2 = target_valid2.sample(n=500, replace=True, random_state=state)
    pred_valid_subsample2 = predictions_valid2[target_valid_sub2.index]
    predictions_valid_sorted2 = pred_valid_subsample2.sort_values(ascending=False)
    values2.append(profit(predictions_valid_sorted2, target_valid_sub2, 200))
    
values2 = pd.Series(values2)
lower2 = values2.quantile(0.025)

mean2 = values2.mean()
print("Средняя прибыль: {:.2f}".format(mean2))
# interval2 = t.interval(0.95, len(values2), values2.mean(), values2.sem())

# print(interval2)
lower2 = values2.quantile(.05)
upper2 = values2.quantile(.95)

print('95% доверительный интервал в регионе (тысяч руб.): {:.2f} - {:.2f}'.format(lower2, upper2))

Средняя прибыль: 408545.68
95% доверительный интервал в регионе (тысяч руб.): -56146.25 - 869515.28


In [70]:
risk_of_loss2 = values0[values2 < 0].count()/1000
print('Риск убытков в регионе 1: {:.2%}'.format(risk_of_loss2))

Риск убытков в регионе 1: 7.50%


In [71]:
results3['Данные предсказаний регион 3'] = ['{:.2f}'.format(mean2), 
                                            '{:.2f} : {:.2f}'.format(lower2, upper2), '{:.2%}'.format(risk_of_loss2)]
results3

Данные предсказаний регион 1  \
Средняя прибыль в регионе после Bootstrap (тыся...                    427847.56   
95% доверительный интервал в регионе (тысяч руб.)          -9708.12 : 865127.64   
Риск убытков в регионе в %                                                5.50%   

                                                   Данные предсказаний регион 2  \
Средняя прибыль в регионе после Bootstrap (тыся...                    511530.22   
95% доверительный интервал в регионе (тысяч руб.)         149485.76 : 862955.17   
Риск убытков в регионе в %                                                0.60%   

                                                   Данные предсказаний регион 3  
Средняя прибыль в регионе после Bootstrap (тыся...                    408545.68  
95% доверительный интервал в регионе (тысяч руб.)         -56146.25 : 869515.28  
Риск убытков в регионе в %                                                7.50%

<div class="alert alert-block alert-success">
<b>Успех:</b> Бутстрап применен корректно. Риски посчитаны верно.
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> 95% дов интервал для этой задачи нужно искать через 0.975 и 0.025 квантили.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> 
Я на самом деле сначала так и написал, а потом меня что-то смутило.. 
    
Я еще почитаю подробнее про эти интервалы, щас как раз прохожу на степике курс по статистике. 
</div>

<div class="alert alert-block alert-success">
<b>Успех[2]: </b>
   
Понял, дополнительный курсы - это замечательно!) Если вкратце, то вызывая st.t.interval(...), ты находишь 95%-доверительный интервал для значения **средней** прибыли, то есть промежуток, в котором с большой вероятностью при проведении аналогичных экспериментов окажется средняя прибыль. По заданию требовалось оценить 95%-доверительный интервал для значения **прибыли**, то есть найти такой промежуток, в котором с большой вероятностью окажется прибыль случайно выбранного месторождения данного региона. Нижняя граница этого интервала – это 2.5%-квантиль полученных значений прибыли, верхняя граница – 97.5%-квантиль.

</div>

## Вывод

Мы брали по 500 случайных скважин каждого региона, далее сортировали их в порядке убывания по количеству сырья, потом отбирали из них 200 лучших, считали среднее значение и записывали его в переменную. Так мы сделали 1000 раз и получили результаты:
1. Средняя прибыль в регионе номер 2 будет самой большой
2. Мы получили 95% доверительный интервал **в этом интервале с 95% вероятностью будет находиться среднее значение прибыли**
3. мы также посчитали % убыточных скважин. 

Исходя из полученных результатов отвечаем на наши главные вопросы: 
1. **Построить модель для определения региона, где добыча принесёт наибольшую прибыль.** - это регион номер 2
2. **Проанализировать возможную прибыль и риски.** - наименьшие риски убыточных скважин - так же 2-й регион

Делаем вывод что бурить нужно во 2-м регионе!

<div class="alert alert-block alert-success">
<b>Успех:</b> Всегда приятно видеть структурированный итоговый вывод в конце проекта.
</div>

# Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: модели обучены и проверены
    - [x]  Данные корректно разбиты на обучающую и валидационную выборки
    - [x]  Модели обучены, предсказания сделаны
    - [x]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [x]  На экране напечатаны результаты
    - [x]  Сделаны выводы
- [x]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [x]  Для всех ключевых значений созданы константы Python
    - [x]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [x]  По предыдущему пункту сделаны выводы
    - [x]  Написана функция расчёта прибыли
- [x]  Выполнен шаг 4: посчитаны риски и прибыль
    - [x]  Проведена процедура *Bootstrap*
    - [x]  Все параметры бутстрепа соответствуют условию
    - [x]  Найдены все нужные величины
    - [x]  Предложен регион для разработки месторождения
    - [x]  Выбор региона обоснован